src/core/ddsi/src/ddsi_raweth.c

```diff
@@ -391,6 +395,334 @@ static int joinleave_asm_mcgroup (ddsrt_socket_t socket, int join, const ddsi_lo
+        if (vtag)
+        {
+          port += ((uint32_t)(vtag->tag & 0xfff) << 20) + ((uint32_t)(vtag->tag & 0xf000) << 16);
+          if (srcloc)
+          {
+            srcloc->kind = DDSI_LOCATOR_KIND_RAWETH;
+            srcloc->port = port;
+            memset(srcloc->address, 0, 10);
+            memcpy(srcloc->address + 10, eth_hdr->smac, 6);
+          }
+        }
+      }
+      else
+      {
+        char addrbuf[DDSI_LOCSTRLEN];
+        (void) snprintf(addrbuf, sizeof(addrbuf), "[%02x:%02x:%02x:%02x:%02x:%02x]:%u",
+                  eth_hdr->smac[0], eth_hdr->smac[1], eth_hdr->smac[2],
+                  eth_hdr->smac[3], eth_hdr->smac[4], eth_hdr->smac[5], vtag ? ntohs(vtag->proto) : ntohs(eth_hdr->proto));
+        DDS_CWARNING(&conn->m_base.gv->logconfig, "%s => %d truncated to %d\n", addrbuf, (int)ret, (int)len);
+        rc = DDS_RETCODE_ERROR;
+        goto error;
+      }
+    }
+    // else drop packet because it was truncated thus exceeded buffer size.
+
+    uc->bptr += BPF_WORDALIGN(bpf_hdr->bh_hdrlen + bpf_hdr->bh_caplen);
+    if (uc->bptr >= uc->buffer + uc->avail)
+      uc->avail = 0;
+  }
+  else
+    uc->avail = 0;
+
+error:
+  ddsrt_mutex_unlock (&uc->lock);
+  return (rc == DDS_RETCODE_OK ? ret : -1);;
+}
+
+static ssize_t ddsi_raweth_conn_write (struct ddsi_tran_conn * conn, const ddsi_locator_t *dst, const ddsi_tran_write_msgfrags_t *msgfrags, uint32_t flags)
+{
+  ddsi_raweth_conn_t uc = (ddsi_raweth_conn_t) conn;
+  dds_return_t rc = DDS_RETCODE_OK;
+  ssize_t ret;
+  struct ddsi_vlan_header vhdr;
+  uint16_t vtag;
+  size_t hdrlen;
+  (void) flags;
+
+  assert(msgfrags->niov <= INT_MAX - 1); // we'll be adding one later on
+ 
+  vtag = (uint16_t)(((dst->port >> 20) & 0xfff) + (((dst->port >> 16) & 0xe) << 12));
+  
+  memcpy(vhdr.e.dmac, dst->address + 10, 6);
+  memcpy(vhdr.e.smac, uc->m_base.m_base.gv->interfaces[0].loc.address + 10, 6);
+  
+  if (vtag) {
+    vhdr.e.proto = htons ((uint16_t) ETHERTYPE_VLAN);
+    vhdr.vtag = htons (vtag);
+    vhdr.proto = htons ((uint16_t) uc->m_base.m_base.m_port);
+    hdrlen = sizeof(vhdr);
+  } else {
+    vhdr.e.proto = htons ((uint16_t) uc->m_base.m_base.m_port);
+    hdrlen = 14;
+  }
+
+  DDSRT_STATIC_ASSERT(DDSI_TRAN_RESERVED_IOV_SLOTS >= 1);
+
+  ddsrt_iovec_t * const iovs = (ddsrt_iovec_t *) &msgfrags->tran_reserved[DDSI_TRAN_RESERVED_IOV_SLOTS - 1];
+  iovs[0].iov_base = &vhdr;
+  iovs[0].iov_len = hdrlen;
+
+  if ((ret = writev (uc->m_sock, iovs, (int)(msgfrags->niov + 1))) < 0)
+  {
+      DDS_CERROR(&conn->m_base.gv->logconfig, "ddsi_raweth_conn_write failed with retcode %"PRIdSIZE, ret);
+      rc = DDS_RETCODE_ERROR;
+  }
+
+  return (rc == DDS_RETCODE_OK ? ret : -1);
+}
+
+static dds_return_t ddsi_raweth_set_filter (struct ddsi_tran_factory * fact, ddsrt_socket_t sock, uint32_t port)
+{
+  int r;
+  ushort etype = (ushort)(port & 0xFFFF);
+  struct bpf_insn insns[] = {
+    BPF_STMT(BPF_LD+BPF_H+BPF_ABS, 12),
```

Comment:
Nice to have a somewhat readable representation compared to Linux. Does this by any chance work on Linux as well?